rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,227 @@
+determine protocol of an exposed port:
+-  if kubernetes service port definition has `appProtocol` field, and it contains
+values `http`/`https` we will use it in the URI.
+- if exposed port's name is `http` or it has numeric value 80, `http` will be used.
+- if exposed port's name is `https` or it has numeric value 443, `https` will be used.
+
+Annotation `teleport.dev/discovery-port` controls preferred port for the Kubernetes service, only this one will be used even if service
+has multiple exposed ports. Its value should be one of the exposed service ports; otherwise, the app will not be imported. 
+Value can be matched either by numeric value or by the name of the port defined on the service.
+
+Annotation `teleport.dev/app-rewrite` controls rewrite configuration for Teleport app, if needed. It should
+contain full rewrite configuration in YAML format, as you would put into `rewrite` config section of an 
+app (see [documentation](https://goteleport.com/docs/application-access/guides/connecting-apps/#rewrite-redirect)).
+```yaml
+annotations:
+  teleport.dev/app-rewrite: |
+    redirect:
+    - "localhost"
+    - "jenkins.internal.dev"
+    headers:
+    - "X-Custom-Header: example"
+    - "Authorization: Bearer {{internal.jwt}}"
+```
+
+
+Example Kubernetes service configs using annotations can look like this:
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: first-service
+  labels:
+    app: first-service
+  annotations:
+    teleport.dev/discovery-type: app # Allowed values are [`app`]
+    teleport.dev/discovery-protocol: http # Allowed values are [`http`, `https`, `tcp`]
+    teleport.dev/discovery-port: 80
+spec:
+  ports:
+  - name: http
+    port: 80
+    protocol: TCP
+    targetPort: 80
+  - name: postgres
+    port: 5432
+    protocol: TCP
+    targetPort: 5432
+  selector:
+    app: first-pod
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: second-service
+  labels:
+    app: second-service
+  annotations:
+    teleport.dev/discovery-protocol: tcp 
+    teleport.dev/discovery-port: fluentd
+spec:
+  ports:
+    - name: http
+      port: 80
+      protocol: TCP
+      targetPort: 80
+    - name: fluentd
+      port: 24224
+      protocol: TCP
+      targetPort: 24224
+  selector:
+    app: second-pod
+```
+
+Discovery for the Kubernetes apps will support dynamic discovery configuration, when it will be implemented - [RFD 125](https://github.com/gravitational/teleport/blob/master/rfd/0125-dynamic-auto-discovery-config.md).
+
+### App service
+
+App service will be running inside the Kubernetes cluster and proxy connection from remote users to local cluster IP. App service
+already has capability to watch dynamically created app resources, users just need to correctly setup
+`resources` field in the config:
+
+```yaml
+app_service:
+  enabled: yes
+  resources:
+    - labels:
+        "teleport.dev/kubernetes-cluster": "main-cluster"
```

Comment:
Like mentioned above, I would just generate a discovery group when installing a Helm chart and then use it in both discovery_service and app_service. You can template it to be the cluster name as well, should be easy to do in the Helm chart, and one less configuration field.

Reply:
We discussed with Tiago and having label `"teleport.dev/kubernetes-cluster"` seems more clear intent. But yeah, we can populate it from discovery group and we can set discovery group to be kube cluster name in the helm chart, so we will be able to drop `kubernetes_cluster` field from discovery service config.